### ISSUE TITLE
[FEATURE] Quantization processor in ingest pipeline

### ISSUE THREAD

**Issue by YeonghyeonKO** [opensearch-project/neural-search#991]
### Is your feature request related to a problem?
After documents are ingested by **text_embedding** processor, an array of float32 type per **knn_vector** field is stored in segments.(hnsw or ivf) 

But when setting index mappings quantized as byte or binary type, documents are no longer able to be embedded just only with **text_embedding** ingest processor. They need to be quantized before that can occur inconvenience to make quantizers somewhere outside of OpenSearch Clusters. 

And vice versa, query vectors must be converted from float to byte type when calculating similarities with "already converted" vector fields in segments. 

Is there any plan to support so called **"quantization"** processor when it comes to "byte" type of knn_vector field?

<br>

---

(except Faiss's SQfp16 and Lucene's scalar quantization which don't require to quantize before indexing)
> At ingestion time, when you upload 32-bit floating-point vectors to OpenSearch, SQfp16 quantizes them into 16-bit floating-point vectors and stores the quantized vectors in a k-NN index. At search time, SQfp16 decodes the vector values back into 32-bit floating-point values for distance computation. ([doc](https://opensearch.org/docs/latest/search-plugins/knn/knn-vector-quantization/#faiss-16-bit-scalar-quantization))

> Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. ([doc](https://opensearch.org/docs/latest/search-plugins/knn/knn-vector-quantization/#lucene-scalar-quantization))

**Comment by heemin32** [comment]
@YeonghyeonKO Thanks for the feature request. 
Is it possible to ingest byte or binary data vectors using an [embedding model](https://cohere.com/blog/int8-binary-embeddings) that supports byte or binary embeddings, or does this remain a challenge even when using the text_embedding processor?

**Comment by YeonghyeonKO** [comment]
@heemin32 
Yes, documents can be easily indexed using int8/binary embedding models(proprietary) like Cohere and it's not suitable for general purpose since they usually require commercial license or type of API key subscription.

But let's say we're in the closed network, developers cannot deploy [pretrained-models](https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/) directly provided by OpenSearch (because a firewall will block communication with `artifacts.opensearch.org`) unless a torch_script.zip has already been downloaded to the local. Further, it's impossible to [connect to externally hosted models](https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/index/) in this case. 

In this case, exporting quantized models to torchscript(.pt and with config files) is the only way we can use? Of course, it would be great if OpenSearch supports more pretrained models returning quantized vectors (int8 or binary).

**Comment by heemin32** [comment]
@YeonghyeonKO Thanks for sharing your use case. 
First, please note that if the model is not trained with int8/binary embeddings in mind, the recall tends to be suboptimal, generally below 0.9.

Here are a few options to consider
**Option without recall degradation**
1. Disk-based vector search: This method performs quantization, oversampling, and rescoring automatically, offering high recall with low memory usage, even if the model isn't specifically trained for int8/binary embeddings. https://opensearch.org/docs/latest/search-plugins/knn/disk-based-vector-search/

2. Pretrained Models for Int8/Binary Embeddings: You can propose support for pretrained models that generate int8/binary embeddings in the ml-common repository.

**Option with recall degradation**
3. Int8/Binary Quantization in k-NN: Currently, the k-NN plugin supports fp16 quantization(fp16 qt maintain high recall though). Extending this to support int8/binary quantization could be another approach. https://opensearch.org/docs/latest/search-plugins/knn/knn-vector-quantization/#faiss-16-bit-scalar-quantization

4. Ingest Processor for Int8/Binary Quantization: As outlined in this GitHub issue, implementing int8/binary quantization within the ingest processor is another option.

Since the processor can only handle a single vector value at a time, it is limited to very basic quantization. I recommend considering options 1, 2, and 3 before deciding if option 4 is still the best choice for your needs.

**Comment by YeonghyeonKO** [comment]
According to my priorities depending on man-hours and learning curve, the option 1 would be nice to try on. 
Since both OpenSearch and Elasticsearch recommends Binary Quantization recently for no degradation of recall, I'll research and optimize it in my use case. Thanks for making it clear 👍 @heemin32

**Comment by navneet1v** [comment]
@YeonghyeonKO having a QP could be an interesting feature. But there is a way to do via MLCommons only. Here is a tutorial for byte quantization: https://github.com/opensearch-project/ml-commons/blob/78a304a95d546db31139a135a1f031890a55ae72/docs/tutorials/semantic_search/semantic_search_with_byte_quantized_vector.md?plain=1#L3

**Comment by YeonghyeonKO** [comment]
@navneet1v This tutorial is truly helpful for developers who can freely use external API(Cohere or OpenAI), but not for who have restrictions in security. Is there any other way to create own connectors using OpenSearch's pretrained Models?

**Comment by dblock** [comment]
[[Catch All Triage](https://github.com/opensearch-project/.github/blob/main/TRIAGING.md) - [1](https://github.com/dblock), [2](https://github.com/reta), [3](https://github.com/andrross/), [4](https://github.com/krisfreedain)]

**Comment by heemin32** [comment]
I don't think the need of quantization processor is very strong. @YeonghyeonKO, olease reopen if you think it is still required feature.
